bundles/org.openhab.binding.fmiweather/src/main/java/org/openhab/binding/fmiweather/internal/AbstractWeatherHandler.java

```diff
@@ -0,0 +1,233 @@
+        }
+        if (data.values[0] == null) {
+            return -1;
+        }
+        for (int i = 1; i < data.values.length; i++) {
+            if (data.values[i] == null) {
+                return i - 1;
+            }
+        }
+        if (data.values[data.values.length - 1] == null) {
+            return -1;
+        }
+        return data.values.length - 1;
+    }
+
+    protected static long floorToEvenMinutes(long epochSeconds, int roundMinutes) {
+        long roundSecs = roundMinutes * 60;
+        return (epochSeconds / roundSecs) * roundSecs;
+    }
+
+    protected static long ceilToEvenMinutes(long epochSeconds, int roundMinutes) {
+        double epochDouble = epochSeconds;
+        long roundSecs = roundMinutes * 60;
+        double roundSecsDouble = (roundMinutes * 60);
+        return (long) Math.ceil(epochDouble / roundSecsDouble) * roundSecs;
+    }
+
+    /**
+     * Update QuantityType channel state
+     *
+     * @param channelUID channel UID
+     * @param epochSecond value to update
+     * @param unit unit associated with the value
+     */
+    protected <T extends Quantity<T>> void updateEpochSecondStateIfLinked(ChannelUID channelUID, long epochSecond) {
+        if (isLinked(channelUID)) {
+            updateState(channelUID, new DateTimeType(ZonedDateTime.ofInstant(Instant.ofEpochSecond(epochSecond), UTC)
+                    .withZoneSameInstant(ZoneId.systemDefault())));
+        }
+    }
+
+    /**
+     * Update QuantityType or DecimalType channel state
+     *
+     * Updates UNDEF state when value is null
+     *
+     * @param channelUID channel UID
+     * @param value value to update
+     * @param unit unit associated with the value
+     */
+    protected void updateStateIfLinked(ChannelUID channelUID, @Nullable BigDecimal value, @Nullable Unit<?> unit) {
+        if (isLinked(channelUID)) {
+            if (value == null) {
+                updateState(channelUID, UnDefType.UNDEF);
+            } else if (unit == null) {
+                updateState(channelUID, new DecimalType(value));
+            } else {
+                updateState(channelUID, new QuantityType<>(value, unit));
+            }
+        }
+    }
+
+    /**
+     * Unwrap optional value and log with ERROR if value is not present
+     *
+     * @param optional optional to unwrap
+     * @param messageIfNotPresent logging message
+     * @param args arguments to logging
+     * @throws IllegalStateException when value is not present
+     * @return unwrapped value of the optional
+     */
+    protected <T> T unwrap(Optional<T> optional, String messageIfNotPresent, Object... args) {
+        if (optional.isPresent()) {
+            return optional.get();
+        } else {
+            // logger.error(messageIfNotPresent, args) avoided due to static analyzer
+            String formattedMessage = String.format(messageIfNotPresent, args);
+            logger.error("Unwrapping error: {}", formattedMessage);
```

Comment:
OFFLINE/ONLINE status was already handled by the catch clause, but I see your point.

Changed unwrap to use FMIUnexpectedResponseException since all failing unwraps are due to unexpected content in the API response. Using warn level.

See a17d51c68be3a756d4de83bb2c1793c5768b9ba6